### ISSUE TITLE
Enhancement: [await-thenable] should check that `for-await` loop is used on an async iterable

### ISSUE THREAD

**Issue by kirkwaiblinger** [typescript-eslint/typescript-eslint#8858]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=5.4.3&showAST=es&fileType=.tsx&code=GYewTgBAhg7lCWAXCAKAxiAdgZ2diIwEA2gOSYikA0EpiAFgKaZQBGANo9tbU2I6QC6ASggBvAL4AoRgA8ADuGRiIEoA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6AQwHd3K78ALRE3YAjJOiiJo0APbRI4MAF8QSoA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false

### Repro Code

```TypeScript
for await (const s of ['no', 'thenables', 'here']) {}
export { }
```

### ESLint Config

```javascript
module.exports = {
  "rules": {
    "@typescript-eslint/await-thenable": "error"
  }
}
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    "strictNullChecks": true
  }
}
```

### Expected Result

I expect that the first line should flag due to `for await` being used on something that does not have a `Symbol.asyncIterator`.

### Actual Result

No error.

### Additional Info

Note that it's unambiguous IMO that the example given should flag. However, I, personally, would also like for the rule to flag on misused for-await-of, where the iterable is a _synchronous_ iterator over promises. That should be written as an ordinary for loop. It's only a true async iterable that should be allowed to be iterated over in a for-await-of loop.

Refresher on for-await-of: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/for-await...of

**Comment by Josh-Cena** [comment]
This sounds a lot more like the promise aggregators rule (https://github.com/typescript-eslint/typescript-eslint/pull/8094) where you also want to check "some" or "all" of them are thenable... Which is why again I'm unhappy with that being a separate rule, because people would naturally expect await-thenable to work on things that get implicitly awaited too!

**Comment by kirkwaiblinger** [comment]
@Josh-Cena +1

**Comment by kirkwaiblinger** [comment]
@Josh-Cena 

Note that there is also another completely redundant check to prohibit awaiting thenables in `return-await`, so [two different rules flag this code for the exact same reason](https://typescript-eslint.io/play/#ts=5.0.4&fileType=.ts&code=IYZwngdgxgBAZgV2gFwJYHsL3egFASgC4YAFAJ3QFtUQBTAHhGTNQgHMA%2BGAbwFgAoGDDK1kCMlmAB3YKmQwARBHTzgMZAAtaEYACMANrQUBuAQF8gA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6aRfWaJ2gQwDufSvnRgA2pETRoAe2iQANFD7whhVAF1F4bHiKlkFanQbM2g4XXwALREz4AjJGMnS5CzboC%2BIb0A&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false).

return-await:
https://github.com/typescript-eslint/typescript-eslint/blob/fde44275117b3a7bbdcaaf267c73986b6637c5f9/packages/eslint-plugin/src/rules/return-await.ts#L195-L223

await-thenable:
https://github.com/typescript-eslint/typescript-eslint/blob/fde44275117b3a7bbdcaaf267c73986b6637c5f9/packages/eslint-plugin/src/rules/await-thenable.ts#L37-L64

I guess that's a separate issue though.

**Comment by Josh-Cena** [comment]
Just realized this request isn't about whether the loop is over `AsyncIterable<T>` where `T` is a promise or not, but whether it's over `Iterable<any>` or `AsyncIterable<any>` 🤦‍♂️ Even if you have `AsyncIterable<number>`, you can't use something other than `for await` to unroll it, so we can't report that.

**Comment by Josh-Cena** [comment]
Slightly awkward is the rule naming. We are not checking for thenables here (and even renaming it to `await-promises` wouldn't help). I'm wondering if we should rename this rule to `no-unnecessary-await`, or just put this in another rule.

**Comment by kirkwaiblinger** [comment]
> Slightly awkward is the rule naming. We are not checking for thenables here (and even renaming it to `await-promises` wouldn't help). I'm wondering if we should rename this rule to `no-unnecessary-await`, or just put this in another rule.

Well, it is a quite bad name even for the current usage, and I think that's worthy of a separate discussion. 

As it is, the rule really is dont-await-non-thenables. And I'd argue that an AsyncIterable is something whose `[Symbol.asyncIterator]()` returns an object with a `.next()` method that is intended to return a Thenable. So the notion of "awaiting thenables" is contained in that IMO.
